### PR TITLE
Add support for Ultralight emulation.

### PR DESCRIPTION
**Pull request by turbocool3r**
This PR adds support for Mifare Ultralight emulation.

Currently it's a bit untested and lacks functionality of newer ultralight versions and NTAGs (e.g. password auth). I'll convert it from a WIP PR when it's ready. In the current state it can emulate the original Ultralight and the UID magic mode works.

### CODE REVIEW

**Comment by xianglin1998** [comment]
Perfect👍

**Comment by turbocool3r** [comment]
I've tested that this works with a proxmark so looks good to me. If someone could test this thoroughly as well it would be nice.

Currently I've tested:
* Password auth for MF0UL11 with various AUTH0 values, enabled/disabled UID mode,
* R/O locking using static lock bytes (both block and page) for MF0ICU1, MF0UL11,
* R/O locking using dynamic lock bytes (both block and page) for MF0UL11,
* CFGLCK bit,
* signature reading, get version, counters for MF0UL11 and MF0UL21. checked that this doesn't work on MF0ICU1.
* invalid commands returning NAK (though I'm not sure which values are to be used in certain cases),
* PWD/PACK reading in non-UID mode returns zeroes, in UID the actual values.

Problems:
* state is always authenticated after the first successful auth. not sure if the reset handler is the right place to change that (though it seems so),
* no mirroring support on NTAG (should I add it rn?)
* maybe something else

**Comment by taichunmin** [comment]
Should we support the NFC counter mirror feature of NTAG?

**Comment by turbocool3r** [comment]
I could add this as well.

**Comment by turbocool3r** [comment]
I've tested NTAG mirroring and it works well. Also fixed bugs related to authentication and counters, mostly NTAG related. Tested that the behavior matches an actual card. Lastly added some more cli commands and improved some of the existing commands (see [this](https://github.com/RfidResearchGroup/ChameleonUltra/pull/215/commits/2da6d35784497742eb29db93d185524db098eaf1) commit).

File: firmware/application/src/rfid/nfctag/tag_emulation.c [hunk 1]
```diff
@@ -71,7 +71,7 @@ static tag_slot_config_t slotConfig ALIGN_U32 = {
     // See tag_emulation_factory_init for actual tag content
     .slots = {
         { .enabled_hf = true,  .enabled_lf = true,  .tag_hf = TAG_TYPE_MIFARE_1024, .tag_lf = TAG_TYPE_EM410X,    },   // 1
-        { .enabled_hf = true,  .enabled_lf = false, .tag_hf = TAG_TYPE_MIFARE_1024, .tag_lf = TAG_TYPE_UNDEFINED, },   // 2
+        { .enabled_hf = true,  .enabled_lf = false, .tag_hf = TAG_TYPE_MF0ICU1,     .tag_lf = TAG_TYPE_UNDEFINED, },   // 2
```
Comment (taichunmin) [inline firmware/application/src/rfid/nfctag/tag_emulation.c:74]: the `.tag_hf` of factory reset data changed?

Comment (turbocool3r) [reply firmware/application/src/rfid/nfctag/tag_emulation.c:74]: I can remove this, I did that for tests initially.

Comment (turbocool3r) [reply firmware/application/src/rfid/nfctag/tag_emulation.c]: Done.

**Comment by taichunmin** [comment]
Run `hf mfu eview` on mifare slot should thow errer.

**Comment by turbocool3r** [comment]
> Run `hf mfu eview` on mifare slot should thow errer.

Fixed in [3ff1c78](https://github.com/RfidResearchGroup/ChameleonUltra/pull/215/commits/3ff1c785a51b8e89a79328ea1c8bca4802ca3226).

File: software/script/chameleon_cli_unit.py [hunk 2]
```diff
@@ -1679,30 +1681,188 @@ def on_exec(self, args: argparse.Namespace):
...
+            'keep_rf_field': 0,
+            'check_response_crc': 1,
+        }
+        
+        if param.pwd is not None:
+            options['keep_rf_field'] = 1
+            resp = self.cmd.hf14a_raw(options=options, resp_timeout_ms=200, data=struct.pack('!B', 0x1B)+param.pwd)
+            options['keep_rf_field'] = 0
+            options['auto_select'] = 0
+            print(f" - PACK: {resp[:2].hex()}")
```
Comment (taichunmin) [inline software/script/chameleon_cli_unit.py:1749]: It should stop write page while auth failed

File: software/script/chameleon_cli_unit.py [hunk 3]
```diff
@@ -1679,30 +1681,188 @@ def on_exec(self, args: argparse.Namespace):
...
             'check_response_crc': 1,
         }
-        # TODO: auth first if a key is given
+
+        if param.pwd is not None:
+            options['keep_rf_field'] = 1
+            resp = self.cmd.hf14a_raw(options=options, resp_timeout_ms=200, data=struct.pack('!B', 0x1B)+param.pwd)
+            options['keep_rf_field'] = 0
+            options['auto_select'] = 0
+            print(f" - PACK: {resp[:2].hex()}")
```
[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py:1690]: Screenshot 2024-07-07 at 19 39 50

it will

[the remark this answers is not in the document]
Comment (taichunmin) [reply software/script/chameleon_cli_unit.py:1690]: Should we check NACK response?
Xnip2024-07-08_00-50-48

[the remark this answers is not in the document]
Comment (taichunmin) [reply software/script/chameleon_cli_unit.py:1699]: Because I think to distinguish not found and wrong auth is important.

[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py:1699]: I'll do. The cards I've tested with seem to just not send the response instead of sending NAKs

[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py:1699]: Hm, that's a bit complicated. The `HF14A_RAW` command only returns the amount of bytes received, NAKs and ACKs get rounded to 1 byte. With this command we could just check if byte is 0xA0 or something else, but there should probably be another command that can return the amount of bits. Also for some reason while testing I've usually noticed an empty response returned instead of one byte

[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py:1699]: I see, that's because when you want CRC calculation you can't receive ACKs and NAKs

[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py]: This should be done now.

Screenshot 2024-07-08 at 01 05 17

[the remark this answers is not in the document]
Comment (taichunmin) [reply software/script/chameleon_cli_unit.py]: how about try to set appendCrc to false and check if resp is `0x0A`?

ref: https://github.com/RfidResearchGroup/ChameleonUltra/blob/3cadd47eebbc719ab97626845373972756e3d80e/software/script/chameleon_cmd.py#L968

[the remark this answers is not in the document]
Comment (turbocool3r) [reply software/script/chameleon_cli_unit.py]: It will be the PACK value instead of 0x0A on success. We can just add another command version that returns size in bits I think if we really want to ensure NAKs are NAKs.

[the remark this answers is not in the document]
Comment (taichunmin) [reply software/script/chameleon_cli_unit.py]: Just a recommendation. In my js sdk, I will set `wait_response` to true, `check_response_crc` to false, then check `pack + crc16` for success, NAK for auth failed (hf14a_raw return 8 bit is OK), and other for unknown response.

[the remark this answers is not in the document]
Comment (taichunmin) [reply software/script/chameleon_cli_unit.py]: NAK is `0x04`, not `0x0A`. My mistake.

**Comment by taichunmin** [comment]
I think the NAK of the NTAG emulation should not only return 0x04.

Xnip2024-07-09_00-26-50

**Comment by taichunmin** [comment]
Should we add cmd to get/set emu's counter?

**Comment by xianglin1998** [comment]
Everything ready?

**Comment by turbocool3r** [comment]
> Everything ready?

I'll do last tests on write locking in the evening and then it should be done.

**Comment by xianglin1998** [comment]
> > Everything ready?
> 
> I'll do last tests on write locking in the evening and then it should be done.

OK, and you can reuse 'STATUS_PAR_ERR'.

**Comment by turbocool3r** [comment]
> I think the NAK of the NTAG emulation should not only return 0x04.

I'll check what the original cards reply. So far Ultralight cards mostly ignore bad commands while NTAGs reply with NAKs.

**Comment by turbocool3r** [comment]
> OK, and you can reuse 'STATUS_PAR_ERR'.

Where exactly do you want me to reuse this status, @xianglin1998? Isn't it more clear what the message is when the statuses are different? Also in some places this helps to understand whether wrong slot type is chosen or arguments are invalid.

**Comment by turbocool3r** [comment]
@taichunmin get/set counter is done, NAKs seem to be fixed and now work like real cards, NTAG counter index fixed.

**Comment by cyber-vi-king** [comment]
Is there support for 7B ULEV1?

pm3 shows such cards as `MIFARE Ultralight EV1 48bytes (MF0UL1101)`

**Comment by turbocool3r** [comment]
> Is there support for 7B ULEV1?
> 
> 
> 
> pm3 shows such cards as `MIFARE Ultralight EV1 48bytes (MF0UL1101)`
> 
> but if I configure a slot to `MF0UL11` it want's 80 bytes and has a 4b UID

Are you sure about the UID size? The code for slot initialization sets 7 byte size explicitly. Did you initialize the slot by doing `hw slot type -s n -t MF0UL11` and later `hw slot init -s n -t MF0UL11` (n is the slot number)?

48 bytes mean the amount of memory usable as general-purpose storage, the tag has more memory than that.

**Comment by turbocool3r** [comment]
I've fixed some bugs and added support for NTAG 210/212. So far all commands seem to be tested. Each MFU compatible tag type is properly detected by proxmark. Controlling counters and version/signature works. Loading and saving dumps from emulator also works. The only thing that is missing is Ultralight C authentication, but that's something that I think can be done in a follow up PR.

**Comment by xianglin1998** [comment]
Perfect!!! Can i merge it?
